In [1]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import RMSprop

In [2]:
# Loading training data
(mnist_train_images, mnist_train_labels), (mnist_test_images, mnist_test_labels) = mnist.load_data()

# Converting data to format expected by Keras #
train_images = mnist_train_images.reshape(60000, 784)
test_images = mnist_test_images.reshape(10000, 784)
# Converting to floating point 32
train_images = train_images.astype('float32')
test_images = test_images.astype('float32')
# Normalizing to 0-1 range, 8-bit integer is 0-255. 
# 0-255/255 leaves us with a range of 0-1.
train_images /= 255 
test_images /= 255

# Converting the resulting labels for the numbers being read in (0-9) to one-hot format
train_labels = keras.utils.to_categorical(mnist_train_labels, 10)
test_labels = keras.utils.to_categorical(mnist_test_labels, 10)

In [3]:
# Creation of a simple sequential model for training.
model = Sequential()
# We will be feeding a 784 feature feed into a RELU layer of 512 nodes
model.add(Dense(512, activation='relu', input_shape=(784,)))
# Output from RELU layer fed into a 10 node softmax layer
model.add(Dense(10, activation='softmax'))

# Setting up model's loss function and optimizer
model.compile(loss='categorical_crossentropy',
              optimizer=RMSprop(),
              metrics=['accuracy'])

Instructions for updating:
Colocations handled automatically by placer.


In [4]:
# Closing out old session if exists, then starting a new session
if 'session' in locals() and session is not None:
    print('Close interactive session')
    session.close()
    
config = tf.ConfigProto()
config.gpu_options.allow_growth=True
sess = tf.Session(config=config)

# Finally, training the model through 10 epochs
history = model.fit(train_images, train_labels,
                    batch_size=100,
                    epochs=10,
                    verbose=2,
                    validation_data=(test_images, test_labels))

Train on 60000 samples, validate on 10000 samples
Instructions for updating:
Use tf.cast instead.
Epoch 1/10
 - 4s - loss: 0.2442 - acc: 0.9298 - val_loss: 0.1388 - val_acc: 0.9580
Epoch 2/10
 - 3s - loss: 0.0995 - acc: 0.9700 - val_loss: 0.0876 - val_acc: 0.9730
Epoch 3/10
 - 4s - loss: 0.0655 - acc: 0.9805 - val_loss: 0.0744 - val_acc: 0.9769
Epoch 4/10
 - 3s - loss: 0.0476 - acc: 0.9858 - val_loss: 0.0661 - val_acc: 0.9796
Epoch 5/10
 - 4s - loss: 0.0350 - acc: 0.9893 - val_loss: 0.0655 - val_acc: 0.9793
Epoch 6/10
 - 3s - loss: 0.0274 - acc: 0.9917 - val_loss: 0.0655 - val_acc: 0.9813
Epoch 7/10
 - 4s - loss: 0.0205 - acc: 0.9940 - val_loss: 0.0719 - val_acc: 0.9801
Epoch 8/10
 - 4s - loss: 0.0158 - acc: 0.9951 - val_loss: 0.0721 - val_acc: 0.9798
Epoch 9/10
 - 4s - loss: 0.0129 - acc: 0.9962 - val_loss: 0.0640 - val_acc: 0.9834
Epoch 10/10
 - 4s - loss: 0.0098 - acc: 0.9968 - val_loss: 0.0724 - val_acc: 0.9817


In [5]:
# Evaluating the overall accuracy of the model
score = model.evaluate(test_images, test_labels, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.07235214983986006
Test accuracy: 0.9817


As seen above, the model has an overall accuracy of ~98% which could be further improved by tweaking the model or training it through more epochs